---
title: "Case Study 2: Cell type composition vs. Gene expression"
author: "Christian Halter"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc_float: true
  BiocStyle::pdf_document: default
editor_options: 
  markdown: 
    wrap: 80
bibliography: references.bib
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'Case_Study_2.html'))})
---

```{r setup, include=FALSE}
# Solution for the "magick package is required to crop" error:
knitr::opts_chunk$set(
  echo = TRUE, # keep showing code chunks
  message = FALSE, 
  warning = FALSE,
  # This is the line that fixes the cropping issue:
  crop = NULL
)
```

# ECODA vs pseudobulk

**ECODA can be more sensitive to detect differences between samples then
pseudobulk**, especially when studying complex conditions and tissues with
high-resolution cell type annotation including numerous low abundant cell
subtypes.

Additionally, **ECODA is much more interpretable** (10-100 cell types) compared to
gene expression (2'000 - 20'000 genes).

In this example, you will see how a dataset without sample separation looks like
and how it compares to a dataset with differentially abundant cell types, in the cell type compositional and gene expressional space.

Also, it will give you a feeling about the sensitivity of cell type composition
based sample separation versus pseudobulk based sample separation.

--------------------------------------------------------------------------------

# R environment

Check package dependencies and install scECODA

```{r}
# if (!requireNamespace("renv")) install.packages("renv")
# library(renv)
# renv::restore()
# renv::install("carmonalab/scECODA")

remotes::install_github("carmonalab/scECODA")
library(scECODA)
```

--------------------------------------------------------------------------------

# Pseudobulk analysis

scECODA provides a convenient function to calculate and visualize sample
pseudobulks from scRNA-seq data, e.g. from a Seurat object.

In this example you will see:

1.  How to calculate pseudobulks
2.  How to plot a PCA from pseudobulks
3.  How pseudobulk compares to cell type compositional analysis (ECODA)
    -   When pseudobulk and cell type composition perform similarly
    -   Pseudobulk falls short of detecting differences between samples if the
        differences are in low abundant cell types

### When pseudobulk and cell type composition perform similarly

We use the Seurat pbmc3k dataset.

```{r message=FALSE, warning=FALSE}
# Loading an example Seurat dataset
library(Seurat)
library(SeuratData)
InstallData("pbmc3k")
data("pbmc3k")
seurat_object <- UpdateSeuratObject(pbmc3k)
```

The cells of this example dataset are annotated but not assigned to samples or
biological conditions. Let us artificially assign them by creating a sample and
condition column in the metadata.

```{r}
n_samples <- 40
n_groups <- 2

seurat_object$sample_id <- ""
seurat_object$sample_id <- rep(x = paste0("sample", seq(n_samples)), length.out = length(seurat_object$sample_id))
seurat_object$condition_id <- ""
seurat_object$condition_id <- c(rep("condition1", n_samples/n_groups),
                      rep("condition2", n_samples/n_groups))
```

As we randomly assigned cells and samples to two arbitrary "conditions", we see
no separation between samples. All separation metrics show low scores around
zero, indicating indeed a random distribution of our samples between conditions.

```{r}
# Create ECODA object
ecoda_object <- create_ecoda_object(
  seurat_object,
  sample_col = "sample_id",
  celltype_col = "seurat_annotations",
  get_pb = TRUE
)

# # Internally this is calculated with this function:
# pb <- calculate_pseudobulk(
#   count_matrix = seurat_object[["RNA"]]$counts,
#   sample_ids = seurat_object@meta.data[[sample_col]]
# )

plot_pca(
  ecoda_object,
  label_col = "condition_id",
  title = "PCA - Cell type composition"
)
plot_pca(
  ecoda_object,
  slot = "pb",
  label_col = "condition_id",
  title = "PCA - Pseudobulk"
)
```

Observe what happens if you introduce a differential abundance of CD8 T and NK
cells between the two groups.

```{r}
# Multiply CD8 T cells and NK cells within "condition2" by 5:
seurat_object_DiffAbu <- seurat_multiply_cells_in_condition(
  seurat_object = seurat_object,
  condition_column = "condition_id",
  target_condition = "condition2",
  celltype_column = "seurat_annotations",
  target_celltypes = c("CD8 T", "NK"),
  multiplier = 5
)

# Verification (Optional):
# Check the cell counts before:
table(seurat_object@meta.data$condition, seurat_object@meta.data$seurat_annotation)
# Check the cell counts after:
table(seurat_object_DiffAbu@meta.data$condition, seurat_object_DiffAbu@meta.data$seurat_annotations)
```

All separation metrics increased and are now close to their maximum value of 1,
indicating separation of the two groups of samples.

```{r}
ecoda_object_DiffAbu <- create_ecoda_object(
  seurat_object_DiffAbu,
  sample_col = "sample_id",
  celltype_col = "seurat_annotations",
  get_pb = TRUE
)
plot_pca(ecoda_object_DiffAbu, label_col = "condition_id", title = "PCA cell type composition", n_ct_show = 2)
plot_pca(ecoda_object_DiffAbu, slot = "pb", label_col = "condition_id", title = "PCA pseudobulk")
```

It can be seen that CD8 T and NK cells are significantly higher in the
condition2. Interestingly, it can also be seen that all the other cell types
seem to be lower abundant, even though their counts were not changed. However,
their relative abundance decreased. Remember that parts of a composition are not
independent but all are part of a total. If you change one part (e.g. relatively
increase), all others change too (e.g. relatively decrease).

```{r}
plot_boxplot(ecoda_object, label_col = "condition_id", title = "CLR Abundance by Cell Type (with Wilcoxon Test)")
plot_boxplot(ecoda_object_DiffAbu, label_col = "condition_id", title = "CLR Abundance by Cell Type (with Wilcoxon Test)")
```

This can also be plotted as stacked bar plots but it is harder to read.

```{r}
# 1. Plotting Summary by condition_id (your original plot)
plot_barplot(ecoda_object, label_col = "condition_id", plot_by = "group", title = "Mean Relative Abundance by Condition")
plot_barplot(ecoda_object_DiffAbu, label_col = "condition_id", plot_by = "group", title = "Mean Relative Abundance by Condition")

# 2. Plotting Each Sample Separately (the new request)
plot_barplot(ecoda_object, label_col = "condition_id",plot_by = "sample", title = "Relative Abundance for Each Sample")
plot_barplot(ecoda_object_DiffAbu, label_col = "condition_id",plot_by = "sample", title = "Relative Abundance for Each Sample")
```

Before you introduced an artificial differential abundance, there was no major
correlation between any of the cell types. Afterwards, however, it can be
clearly seen that CD8 T and NK cells strongly correlate.

```{r}
plot_corr(ecoda_object)
plot_corr(ecoda_object_DiffAbu)
```

### ECODA outperforms pseudobulk in detecting changes in low abundant cell subtypes

Pseudobulk falls short of detecting differences between samples if the
differences are in very low abundant cell types

To showcase this effect we can again use our semi-synthetic dataset:

First, we make the Seurat object so that it has a lot more cells. For this
purpose, we multiply the cells in all samples (both conditions) by 10. Next,

```{r}
all_celltypes <- colnames(ecoda_object@counts)
low_abundant_celltypes <- c("FCGR3A+ Mono", "B")
high_abundant_celltypes <- all_celltypes[!all_celltypes %in% low_abundant_celltypes]

# Let's remove the platelets and DCs as they are zero in about half the samples
# They would introduce noise which would be further amplified when multiplying
seurat_object_clean <- seurat_object
seurat_object_clean$seurat_annotations[seurat_object_clean$seurat_annotations %in% c("Platelet", "DC")] <- NA

# We multiply most cell types by 10 in both conditions:
seurat_object_DiffAbu <- seurat_multiply_cells_in_condition(
  seurat_object = seurat_object_clean,
  condition_column = "condition_id",
  target_condition = "condition1",
  celltype_column = "seurat_annotations",
  target_celltypes = high_abundant_celltypes,
  multiplier = 10
)
seurat_object_DiffAbu <- seurat_multiply_cells_in_condition(
  seurat_object = seurat_object_DiffAbu,
  condition_column = "condition_id",
  target_condition = "condition2",
  celltype_column = "seurat_annotations",
  target_celltypes = high_abundant_celltypes,
  multiplier = 10
)

# Next, we introduce a differential abundance in the low abundant cell types
seurat_object_DiffAbu <- seurat_multiply_cells_in_condition(
  seurat_object = seurat_object_DiffAbu,
  condition_column = "condition_id",
  target_condition = "condition2",
  celltype_column = "seurat_annotations",
  target_celltypes = low_abundant_celltypes,
  multiplier = 5
)


# Verification (Optional):
# Check the cell counts before:
table(seurat_object@meta.data$condition, seurat_object@meta.data$seurat_annotation)
# Check the cell counts after:
table(seurat_object_DiffAbu@meta.data$condition, seurat_object_DiffAbu@meta.data$seurat_annotations)
```

While cell type composition is very sensitive to pick up differential abundances
even in low abundant cell types, pseudobulk gene expression completely fails to
pick up gene expressional differences in low abundant cell types:

```{r}
ecoda_object_DiffAbu <- create_ecoda_object(
  seurat_object_DiffAbu,
  sample_col = "sample_id",
  celltype_col = "seurat_annotations",
  get_pb = TRUE
)
plot_pca(ecoda_object_DiffAbu, label_col = "condition_id", title = "PCA cell type composition", n_ct_show = 4)
plot_pca(ecoda_object_DiffAbu, slot = "pb", label_col = "condition_id", title = "PCA pseudobulk")
```

# References

# Session Info

```{r}
sessionInfo()
```
